feat: Add CapacityBuffers API support (#9276)
aws/karpenter-provider-aws: website/content/en/preview/concepts/capacitybuffers.md

---
title: "CapacityBuffers"
linkTitle: "CapacityBuffers"
weight: 50
description: >
  Understand CapacityBuffers and how they enable pre-provisioned spare capacity for reduced pod scheduling latency.
---

<i class="fa-solid fa-circle-info"></i> <b>Feature State: </b> [Alpha]({{<ref "../reference/settings#feature-gates" >}})

> Karpenter uses CapacityBuffers to pre-provision spare node capacity so that workloads can schedule instantly without waiting for new nodes to launch.

Reviewer: I'd rather reframe this as if we're talking to the user. 'Capacity buffers can be used to pre-provision spare capacity....'

        image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
        resources:
          requests:
            cpu: "2"
            memory: "4Gi"
    nodeSelector:
      karpenter.sh/capacity-type: on-demand
    tolerations:
      - key: workload-type
        value: web
        effect: NoSchedule
```

{{% alert title="Note" color="primary" %}}
PVC-backed and ephemeral volumes in the PodTemplate are automatically stripped from virtual pods since no real PVC exists for topology resolution.
{{% /alert %}}

## spec.scalableRef

Reference to a workload with a scale subresource. Mutually exclusive with `podTemplateRef`. The buffer uses the workload's pod template spec directly and can scale proportionally using `percentage`.

Supported kinds:
- `apps/v1/Deployment`
- `apps/v1/StatefulSet`
- `apps/v1/ReplicaSet`

```yaml
spec:
  scalableRef:
    apiGroup: apps
    kind: Deployment
    name: api-service
  percentage: 20
```

{{% alert title="Note" color="primary" %}}
The buffer reads the workload's `spec.template.spec` directly — no running pods are required for initialization. Changes to the workload spec are picked up within 30 seconds.
{{% /alert %}}

## spec.replicas

> Fixed number of buffer chunks to provision. When used with `percentage`, the minimum of the two is taken. When used with `limits`, the minimum of all constraints is taken.

Reviewer: ```suggestion
Fixed number of buffer chunks to provision. When used with `percentage` or 'limits', the minimum is taken
```

For example, if a Deployment has 10 replicas and `percentage` is 20, the buffer maintains 2 chunks.

## spec.limits

Resource constraints that cap the number of buffer chunks based on total resource requests. If no other constraints are set (`replicas` and `percentage` are both absent), limits determines how many chunks are created. When combined with other constraints, the minimum across all is used.

```yaml
spec:
  podTemplateRef:
    name: worker-template
  limits:
    cpu: "20"
    memory: "40Gi"
```

If each pod requests 2 CPU and 4Gi memory, this produces min(20/2, 40/4) = 10 buffer chunks.

## Replica Calculation

When multiple constraints are specified, Karpenter uses the minimum:

| Configuration | Result |
|---|---|
| `replicas: 5` only | 5 |
| `percentage: 20` with 10-replica Deployment | 2 |
| `replicas: 10` + `limits: {cpu: 3}` (1 CPU per pod) | 3 (min of 10, 3) |
| `replicas: 5` + `percentage: 20` (10-replica Deployment) | 2 (min of 5, 2) |
| `limits: {cpu: 5}` only (1 CPU per pod) | 5 |

## Integration with Disruption

CapacityBuffers integrate with Karpenter's disruption system to ensure buffer capacity is preserved:

> * **Empty consolidation** is blocked for nodes hosting buffer capacity. Even though these nodes have no real pods, the virtual buffer pods prevent them from being treated as empty.

Reviewer: Mention the updated Reason in the condition here I think

* **Underutilized consolidation** is allowed. The replacement node must still fit both real pods and virtual buffer pods, ensuring the buffer is maintained after consolidation.
* **Drift** is allowed. When a NodePool's requirements change and nodes are drifted, Karpenter replaces them and the buffer automatically refills on the new capacity.
* **Expiry** is allowed. After expired nodes are terminated, the buffer controller triggers a new provisioning pass to restore capacity.

> * **Buffer deletion** removes all protection. Nodes that previously hosted buffer capacity become truly empty and are consolidated normally.

Reviewer: Maybe we can cut this?

## Status and Observability

CapacityBuffers include status conditions to help you understand their current state and troubleshoot issues.

### Status Conditions

* **ReadyForProvisioning=True**: Pod template was successfully resolved and target replicas were calculated
* **ReadyForProvisioning=False**: Resolution failed — check the reason for details (PodTemplateNotFound, ScalableRefNotFound, ResolutionFailed)
* **Provisioning=True**: All virtual pods fit on existing cluster capacity without requiring new nodes
* **Provisioning=False**: Virtual pods require new capacity — provisioning is in progress or constrained by NodePool limits

### Example Status

```yaml
status:
  conditions:
    - type: ReadyForProvisioning
      status: "True"
      reason: Resolved
      message: "Pod template resolved successfully"
    - type: Provisioning
      status: "True"
      reason: FitsExistingCapacity
      message: "All 5 virtual pods fit on existing capacity"
  replicas: 5
  podTemplateRef:
    name: web-buffer-template
  podTemplateGeneration: 3
  provisioningStrategy: "buffer.x-k8s.io/active-capacity"
```

> ### Monitoring

Reviewer: I think this should be about the error metrics to watch instead of the statuses. We can also reference the operator_* style metrics once we emit them here

Author: I can make the doc change in the PR that adds operator style metrics
